MazerPlatformer/Character.cs

```diff
@@ -103,30 +110,30 @@ public void MoveInDirection(CharacterDirection direction, GameTime dt)
                     X += MoveByStep();
                     break;
                 default:
-                    throw new ArgumentOutOfRangeException(nameof(direction), direction, null);
+                    return new InvalidDirectionFailure(direction);
             }
-            SetCharacterDirection(direction);
+            return SetCharacterDirection(direction);
         }
 
-        private void SetState(CharacterStates state)
+        private Either<IFailure, Unit> SetState(CharacterStates state)
         {
             CurrentState = state;
-            OnStateChanged?.Invoke(state);
+            return Ensure( () => OnStateChanged?.Invoke(state));
```

Comment:
This is an example of using an external library or component that returns a void. We have no control over how that function will error and throw exceptions. The best we can do is suppress all exceptions through Ensure() can convert them to failures that we can either handle locally or later in the pipeline.
- Question: Should we be handling failures as-soon as possible within the function that they occur in? It does isolate the failure and the handling to the function the failure is presented at...